File: src/service/charactersortingservice.ts
import Character from '../models/character';
import Dungeon from '../models/dungeon';

export default class CharacterSortingService {
  
  public static comparingName(f: Character, s: Character, asc: boolean): number {
    return (f.Name == s.Name ? 0 : (f.Name > s.Name ? 1 : -1)) * (asc ? 1 : -1);
  }

  public static comparingILvl(f: Character, s: Character, asc: boolean): number {
    return (f.ItemLevel - s.ItemLevel) * (asc ? 1 : -1);
  }

  public static comparingRaidProgress(f: Character, s: Character, asc: boolean): number {
    const fDifficult = f.RaidProgress[f.RaidProgress.length - 1];
    const sDifficult = s.RaidProgress[s.RaidProgress.length - 1];

    if (fDifficult == sDifficult) {
      const fProgress = f.RaidProgress.substr(0, f.RaidProgress.indexOf('/'));
      const sProgress = s.RaidProgress.substr(0, s.RaidProgress.indexOf('/'));

      return parseInt(fProgress) - parseInt(sProgress);
    }

    if ((fDifficult == 'M' && sDifficult == 'H') || (fDifficult == 'H' && sDifficult == 'N'))
      return asc ? 1 : -1;

    if ((fDifficult == 'H' && sDifficult == 'M') || (fDifficult == 'N' && sDifficult == 'H'))
      return asc ? -1 : 1;

    return 0;
  }

  public static comparingKeyProgressTank(f: Character, s: Character, asc: boolean): number {
    if (f.ScoreTank == null && s.ScoreTank == null)
      return 0;

    if (f.ScoreTank == null)
      return 1;

    if (s.ScoreTank == null)
      return -1;

    return (f.ScoreTank - s.ScoreTank) * (asc ? 1 : -1);;
  }

  public static comparingKeyProgressHeal(f: Character, s: Character, asc: boolean): number {
    if (f.ScoreHealer == null && s.ScoreHealer == null)
      return 0;

    if (f.ScoreHealer == null)
      return 1;

    if (s.ScoreHealer == null)
      return -1;

    return (f.ScoreHealer - s.ScoreHealer) * (asc ? 1 : -1);;
  }

  public static comparingKeyProgressDps(f: Character, s: Character, asc: boolean): number {
    return (f.ScoreDps - s.ScoreDps) * (asc ? 1 : -1);
  }

  public static comparingKeyProgressAll(f: Character, s: Character, asc: boolean): number {
    return (f.ScoreAll - s.ScoreAll) * (asc ? 1 : -1);
  }

  private static comparingKey(f: Dungeon | null, s: Dungeon | null, asc: boolean): number {
    if (f == null && s == null)
      return 0;

    if (f == null)
      return 1;

    if (s == null)
      return -1;

    return Dungeon.comparingKey(f, s) * (asc ? 1 : -1);
  }

  public static comparingMaxWeekKey(f: Character, s: Character, asc: boolean): number {
    return CharacterSortingService.comparingKey(f.MaxWeekKey, s.MaxWeekKey, asc);
  }

  public static comparingMaxSeasonKey(f: Character, s: Character, asc: boolean): number {
    return CharacterSortingService.comparingKey(f.MaxSeasonKey, s.MaxSeasonKey, asc);
  }

  public static comparingMainAlt(f: Character, s: Character, asc: boolean): number {
    if ((f.MainId == f.Id && s.MainId == s.Id) || (f.MainId != f.Id && s.MainId != s.Id))
      return 0;
    
    if (f.MainId == f.Id)
      return asc ? 1 : -1;

    return asc ? -1 : 1;
  }
}

File: RuthlessBoreas/Services/CharacterSortingService.cs
namespace RuthlessBoreas.Services
{
  using RuthlessBoreas.Models;
  using System;
  using System.Linq;

  public static class CharacterSortingService
  {
    private static bool ascSorting = false;

    public static int SortColumnNumber { get; private set; }

    private static int ComparingILvl(Character c)
    {
      return c.RaiderIo.ItemLevel * (ascSorting ? 1 : -1);
    }

    private static int ComparingRaidProgress(Character c)
    {
      var difficult = c.RaiderIo.RaidProgress[c.RaiderIo.RaidProgress.Length - 1];
      var progress = int.Parse(c.RaiderIo.RaidProgress.Substring(0, c.RaiderIo.RaidProgress.IndexOf('/')));
      var difficultNum = difficult == 'M' ? 300 : (difficult == 'H' ? 200 : 100);

      return (difficultNum + progress) * (ascSorting ? 1 : -1);
    }

    private static int ComparingKeyProgressTank(Character c)
    {
      int result;
      if (c.RaiderIo.ScoreTank.HasValue)
        result = c.RaiderIo.ScoreTank.Value * (ascSorting ? 1 : -1);
      else
        result = 100000;

      return result;
    }

    private static int ComparingKeyProgressHeal(Character c)
    {
      int result;
      if (c.RaiderIo.ScoreHealer.HasValue)
        result = c.RaiderIo.ScoreHealer.Value * (ascSorting ? 1 : -1);
      else
        result = 100000;

      return result;
    }

    private static int ComparingKeyProgressDps(Character c)
    {
      return c.RaiderIo.ScoreDps * (ascSorting ? 1 : -1);
    }

    private static int ComparingKeyProgressAll(Character c)
    {
      return c.RaiderIo.ScoreAll * (ascSorting ? 1 : -1);
    }

    private static int ComparingKey(Dungeon d)
    {
      if (d == null)
        return 0;

      return Dungeon.ComparingKey(d) * (ascSorting ? 1 : -1);
    }

    private static int ComparingMaxWeekKey(Character c)
    {
      return ComparingKey(c.RaiderIo.MaxWeekKey);
    }

    private static int ComparingMaxSeasonKey(Character c)
    {
      return ComparingKey(c.RaiderIo.MaxSeasonKey);
    }

    private static int ComparingMainAlt(Character c)
    {
      return (c.Main.Id == c.Id ? 1 : -1) * (ascSorting ? 1 : -1);
    }

    private static void ToggleSelectedColumn(int column)
    {
      if (column == SortColumnNumber)
      {
        ascSorting = !ascSorting;
      }
      else
      {
        SortColumnNumber = column;
        ascSorting = false;
      }
    }

    public static void OnTableHeaderClickHandler(int column)
    {
      ToggleSelectedColumn(column);
      Func<Character, int> sortingFunction = null;
      switch(column)
      {
        case 1:
          if (ascSorting)
            StorageService.Characters = StorageService.Characters.OrderBy(character => character.Name);
          else
            StorageService.Characters = StorageService.Characters.OrderByDescending(character => character.Name);
          return;
        case 2:
          sortingFunction = ComparingILvl;
          break;
        case 3:
          sortingFunction = ComparingRaidProgress;
          break;
        case 4:
          sortingFunction = ComparingKeyProgressTank;
          break;
        case 5:
          sortingFunction = ComparingKeyProgressHeal;
          break;
        case 6:
          sortingFunction = ComparingKeyProgressDps;
          break;
        case 7:
          sortingFunction = ComparingKeyProgressAll;
          break;
        case 8:
          sortingFunction = ComparingMaxWeekKey;
          break;
        case 9:
          sortingFunction = ComparingMaxSeasonKey;
          break;
      }

      StorageService.Characters = StorageService.Characters.OrderBy(character => sortingFunction?.Invoke(character));
    }
  }
}
